### PR TITLE
feat: Size operator

### LINKED ISSUE #29: [Task] size operator
## What should be done 
$size is a value operator so we need:

### Grammar
- uncomment $size in value operator

### Exec
- add "$size" case in value-operator exec?
  - checks if the element is an array and if the array size has the specified size

### PR DESCRIPTION
**Pull request by jonas-martinez**
Closes #29 

## Checklist
- [x] I didn't over-scope my PR
- [x] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I did not include breaking changes
- [x] I made my own code-review before requesting one

### I included unit tests that cover my changes
  - [x] 👍 yes
  - [ ] 🙅 no, because they aren't needed
  - [ ] 🙋 no, because I need help

### I added/updated the documentation about my changes

- [ ] 📜 README.md
- [ ] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [x] 🙅 no documentation needed

### CODE REVIEW

**Comment by jonas-martinez** [comment]
@pichoemr Can you resolve the conflicts please?

File: test/exec_test.exs [hunk 1]
```diff
@@ -382,5 +382,10 @@ defmodule QueryParser.ExecTest do
                %{"_id" => 10}
              ] = parse_and_exec(data(), %{"_id" => %{"$not" => %{"$lt" => 5}}})
     end
+
+    test "size operator can only take array" do
+      assert_raise ArgumentError, fn -> parse_and_exec(data(), %{"_id" => %{"$size" => 5}}) end
```
Comment (taorepoara) [inline test/exec_test.exs]: I've tested the $size operator on other fields than array and the data is just filtered just like if it returs 0.

File: lib/exec.ex [hunk 2]
```diff
@@ -151,7 +155,10 @@ defmodule QueryParser.Exec do
       "$exists" ->
         nil? = nil == elem_value
         exec_value(value, elem, ctx) != nil?
-
+        
+      "$size" ->
+        elem_value |> length() == exec_value(value, elem, ctx)
```
Comment (pichoemr) [inline lib/exec.ex]:
```suggestion
      "$size" ->
        if is_list(elem_value) do
            elem_value |> length() == exec_value(value, elem, ctx)
        else
            false
        end
```

**Review by pichoemr — CHANGES_REQUESTED** [review]

**Review by taorepoara — APPROVED** [review]

**Review by taorepoara — APPROVED** [review]

File: lib/exec.ex [hunk 3]
```diff
@@ -128,33 +134,7 @@ defmodule QueryParser.Exec do
...
-        elem_value >= exec_value(value, elem, ctx)
-
-      "$exists" ->
-        nil? = nil == elem_value
-        exec_value(value, elem, ctx) != nil?
-
-      "$type" ->
-        is_bson_type(elem_value, exec_value(value, elem, ctx))
-    end
+    exec_value_operator?(operator, value, elem_value, elem, ctx)
```
Comment (pichoemr) [inline lib/exec.ex]:
```suggestion
    exec_value_operator?(operator, elem_value, exec_value(value, elem, ctx))
```

File: lib/exec.ex [hunk 4]
```diff
@@ -204,4 +184,50 @@ defmodule QueryParser.Exec do
       end
     )
   end
+
+  defp exec_value_operator?("$eq", value, elem_value, elem, ctx) do
+    elem_value == exec_value(value, elem, ctx)
+  end
```
Comment (pichoemr) [inline lib/exec.ex]:
```suggestion
  defp exec_value_operator?("$eq",  elem_value, value) do
    elem_value == value  end
```
Same on other exec_value_operator?

File: lib/exec.ex [hunk 5]
```diff
@@ -204,4 +184,50 @@ defmodule QueryParser.Exec do
...
+    exec_value(value, elem, ctx) != nil?
+  end
+
+  defp exec_value_operator?("$size", value, elem_value, elem, ctx) do
+    if is_list(elem_value) do
+      elem_value |> length() == exec_value(value, elem, ctx)
+    else
+      false
+    end
+  end
```
Comment (pichoemr) [inline lib/exec.ex]:
```suggestion
  defp exec_value_operator?("$size", value, elem_value, elem, ctx) when is_list(elem_value) do
      elem_value |> length() == exec_value(value, elem, ctx)
end
  
defp exec_value_operator?("$size", value, elem_value, elem, ctx) do
      false
end
```
Pattern matching is more efficient than if, it's a minor optimization but in this case, it doesn't cost much to do it.

**Review by pichoemr — CHANGES_REQUESTED** [review]

**Review by pichoemr — APPROVED** [review]

**Pull request merged**